Implement configurable escape key for attach/exec

vdemeester opened the pull request:
Implement configurable escape keys (for `attach`, `exec`, `run` and `start`) using the client-side configuration and/or flags `--detach-keys`. This takes some pieces from #6460 from @vieux — except that this time it uses the client configuration (`~/.docker/config.json` for example). 🐧

> This PR allows you to change the set of keys you have to type to detach in TTY mode, it can be less or more that two.
> The basic ASCII table is supported, for example you can do: docker -d --detach-keys="ctrl-@" and detach in one keystroke or docker -d --detach-keys="DEL,+" if you don't like to have a ctrl key

The configuration looks like this :

``` json
{
    "detachKeys": "ctrl-a,a"
}
```

The default keybinding stays the same : `ctrl-p ctrl-q` but it makes possible to configure something else.

There is room for improvements on the code that should be address during the design review process I think :
- For `attach` command I use _queryString_ (`detachKeys=ctrl-a%2Ca`) but for `exec` command I used `runConfig.ExecConfig`. That felt weird and it's not coherent, it should be the same thing for both. It could maybe be passed as HTTP Header, I don't know :sweat_smile: — that's why I'm putting this in review.
- Right now, I don't validate the configuration client side, so if there is something wrong with the `detachKeys` passed, It's just logged and the defaults one are used.
- Some refactoring is there is a little too much duplicated code for my taste (the `keys, err = term.ToBytes(detachKeys)` part is repeated few times..).
- Probably add a `json:omitempty` for `DetachKeys` on `runconfig.ExecConfig`.

What is still to do :
- [X] Rebase it :sweat_smile:
- [X] Implement flags by commands (`run`, `start`, `attach`, `exec`)
- [X] Write some integration tests.
- [X] Update some documentation.
- [X] Update the commit message to make it more clear :sweat_smile:.

🐸

Fixes #3519.

Signed-off-by: Vincent Demeester vincent@sbr.pm

crquan commented:
:+1:

tiborvass commented:
A question to answer is if there are already configs in the client config.json that have no flag equivalents, and if not, then is that something we want ? IOW, if we add a config in config.json, should we also add a flag?

vdemeester commented:
@tiborvass good question :cat: I'm wondering what should be the flag then :sweat_smile:. 

/cc @duglin @thaJeztah @runcom :stuck_out_tongue_closed_eyes:

runcom commented:
> if we add a config in config.json, should we also add a flag

if I'm understanding correctly your question, another question is, what's the use case for having this defined as a flag? do really ppl run their containers assigning different escape keys? why would they?
`docker run -ti --escape-keys "ctrl-a,a" my_cont0 bash`
`docker exec -ti --escape-keys "ctrl-b,b" my_cont1 bash`
I don't like it, but maybe I don't see the value in this. So, I think even if config.json doesn't have configurations w/o an equivalent flag I still see this option better suited in the json instead of a command flag

(notificed the original issue is for `run` command as well)

vdemeester commented:
@runcom talking with @tiborvass a bit on IRC, there might be usecase for a *one-off* different escape keys : when doing *dind* and need to detach from the inner container without detaching from the outer one — editing the config file inside the outer container is not that convenient. A flag would be handy there :sweat_smile:

runcom commented:
Ok that's the use case I haven't thought about :)

tiborvass commented:
As an addendum, I don't think this should be on `exec`, only `run` and possibly be `attach`. I wonder about `start` too.

vdemeester commented:
@tiborvass you're talking only about the `flag` right ? (for not having it on `exec`)

tiborvass commented:
@vdemeester i just realized we still allow detaching from `exec`, that's a mistake. I don't know what to do then.

vdemeester commented:
@tiborvass hum you think we should disable detaching from exec no matter this PR ? I kinda agree on this but pretty sure this should be another PR (probably to be merged before that one). I can take a look :wink:

tiborvass commented:
@vdemeester I agree that issue is a separate one :)

rhatdan commented:
I would want to change this to be system default.  
Is there an easy way to cause the default config.json to have different defaults?

vdemeester commented:
@rhatdan what do you mean by that ?

rhatdan commented:
I would want to change all containers to run with a different detach command.  Something like what tmux does.  I want to set it in a config or on the daemon and be done with it.  I don't want to think about if for every docker run command.

http://superuser.com/questions/74492/whats-the-best-prefix-escape-sequence-for-screen-or-tmux

vdemeester commented:
@rhatdan well this is the goal of this PR : to add a config options (on client side, on `~/.docker/config.json`) so that by default it uses something else, and an option (yet to be written/updated) to be able to do it *per containers*. So this would answer your use case then (the config part of it)

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

LK4D4 commented:
Design makes sense for me, at least daemon-wide.
ping @vdemeester @docker/core-maintainers

vdemeester commented:
Remembering myself I need to rebase this :sweat_smile: :stuck_out_tongue_closed_eyes:

icecrime commented:
I never felt a strong need for this, but it makes sense.

Design LGTM. Tentatively moving to `2-code-review`, @vdemeester can you please rebase?

vdemeester commented:
@icecrime I can yeah `\o/`

crquan commented:
@icecrime I do feel a strong need for this, and the default double Ctrl-P for a previous command in Bash is so bad that bothers me every single day, maybe since day1 of docker.
And it bothers all the ones put their +1 on issue #3519.

icecrime commented:
@crquan Makes sense :-) Let's do it.

calavera commented on daemon/exec.go:
```diff
@@ -165,6 +167,15 @@ func (d *Daemon) ContainerExecCreate(config *runconfig.ExecConfig) (string, erro
+		keys, err = term.ToBytes(config.EscapeKeys)
+		if err != nil {
+			logrus.Warnf("Wrong escape keys provided (%s) using default : ctrl-p ctrl-q", config.EscapeKeys)
+			// return err
```
can we remove the comment if we're not going to return?

vdemeester replied:
oh :open_mouth: Yeah, don't know why it's there :sweat:

thaJeztah commented:
ping @vdemeester ^^ https://github.com/docker/docker/pull/15666#discussion_r43670528

vdemeester commented:
Updated with integration tests, a better commit message and a tiny tiny update on docs (man pages for now).

crquan commented:
sigh, it missed v1.9.0 release; would you @icecrime promise will this awesome work from @vdemeester be in v1.10.0 ?

icecrime commented:
![](http://33.media.tumblr.com/969ecabc49175dbb67ed24cd8cca3214/tumblr_mvkn10eSmO1swukzbo1_400.gif)

tiborvass commented on api/client/attach.go:
```diff
@@ -20,6 +20,7 @@ func (cli *DockerCli) CmdAttach(args ...string) error {
 	cmd := Cli.Subcmd("attach", []string{"CONTAINER"}, Cli.DockerCommands["attach"].Description, true)
 	noStdin := cmd.Bool([]string{"#nostdin", "-no-stdin"}, false, "Do not attach STDIN")
 	proxy := cmd.Bool([]string{"#sig-proxy", "-sig-proxy"}, true, "Proxy all received signals to the process")
+	escapeKeys := cmd.String([]string{"-escapeKeys"}, "", "Set up escape key sequence")
```
why escape keys? why not detach keys? Btw, for flags we don't do camelcase, there should be a hyphen

vdemeester replied:
You're right on both, `--detch-keys` is better, will update :sweat_smile:

tiborvass commented on api/client/exec.go:
```diff
@@ -17,6 +17,7 @@ import (
 // Usage: docker exec [OPTIONS] CONTAINER COMMAND [ARG...]
 func (cli *DockerCli) CmdExec(args ...string) error {
 	cmd := Cli.Subcmd("exec", []string{"CONTAINER COMMAND [ARG...]"}, Cli.DockerCommands["exec"].Description, true)
+	escapeKeys := cmd.String([]string{"-escapeKeys"}, "", "Set up escape key sequence")
```
@vdemeester so in the end, do we have a PR that removes detach from exec? :D

vdemeester replied:
Yeah soon soon :sweat_smile: I'm adding the flag there for consistency now but a PR will be there soon :wink:

vdemeester replied:
@tiborvass the thing is so, we can *only* deprecate it for now right ? (gonna do a PR on deprecating it and we'll see)

tiborvass replied:
i think it goes hand in hand... and yeah i agree it's going to be painful... such a mistake :( Anyway, you don't need to change it in this PR

vdemeester commented:
@tiborvass Updated :wink:

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -134,6 +135,158 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
+	}
+
+	// escape sequence
+	if _, err := cpty.Write([]byte{1}); err != nil {
```
maybe set these byte values to a var

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -134,6 +135,158 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
+	}
+
+	// escape sequence
+	if _, err := cpty.Write([]byte{1}); err != nil {
```
same as above

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -134,6 +135,158 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
+
+	select {
+	case <-ch:
+	case <-time.After(10 * time.Millisecond):
```
Is there a reason the timeout is so short? It's ok to take longer in failure cases.

vdemeester replied:
hum, that was setup like that in `TestRunAttachDetach` so I did the same :wink:

aaronlehmann replied:
I agree, the timeout should be significantly longer. Whenever we have short timeouts like this, we have spurious test failures.

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -134,6 +135,158 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
+	}
+
+	go func() {
+		exec.Command(dockerBinary, "kill", name).Run()
```
Shouldn't the cmd return without this kill call?

vdemeester replied:
@cpuguy83 nop, we're just *detaching* from the container but you're right.. hum.. :sweat_smile:

cpuguy83 replied:
Really I think this invalidates the detach.

cpuguy83 commented on man/docker-attach.1.md:
```diff
@@ -18,15 +19,19 @@ interactively.  You can attach to the same contained process multiple times
-exit code will be returned to the client.
+You can detach from the container (and leave it running) with the
+escape key sequence, default to `CTRL-p CTRL-q`, (for a quiet exit) or
+`CTRL-c` which will send a `SIGKILL` to the container. When you are
```
Really this is `SIGINT`

cpuguy83 commented:
Can you provide some example uses in the docs, both the the CLI flag and the config file?

cpuguy83 commented:
Can we also test the CLI flags trumps the config file (should it?)?

MHBauer commented on api/server/router/local/container.go:
```diff
@@ -419,14 +420,25 @@ func (s *router) postContainersAttach(ctx context.Context, w http.ResponseWriter
+	if detachKeys != "" {
+		keys, err = term.ToBytes(detachKeys)
+		if err != nil {
+			logrus.Warnf("Wrong escape keys provided (%s) using default : ctrl-p ctrl-q", detachKeys)
```
Check my understanding, this checks the string that is coming in can be parsed as key codes.

If that is the true case, I would suggest that the text would be more accurate as "Invalid escape keys provided". My point is, it is not the keys that are wrong, it is the incoming string.

vdemeester replied:
Ah true :wink: :+1:

vdemeester commented:
> Can you provide some example uses in the docs, both the the CLI flag and the config file?

@cpuguy83 I can, I was just wondering where. @thaJeztah wdyt ?

thaJeztah commented:
Haven't looked at the docs yet, but there's a section about client config

Perhaps an example in the `docker attach` docs? E.g. the emacs use case (I think CTRL-p conflicted with emacs?)

crquan commented:
> Perhaps an example in the docker attach docs? E.g. the emacs use case (I think CTRL-p conflicted with emacs?)
Ctrl-P conflicts with Bash, the mostly often used shell, right? double Ctrl-P for a previous command is not good

MHBauer commented:
Control-p is is 'go to previous line' in default emacs. Certainly annoying if you don't use the arrow keys. Similar to the mapping of 'k' in vi.

It would be nice to be able to set the key sequence as per screen or tmux, etc.

thaJeztah commented:
@vdemeester looks like this needs another rebase :cry:

duglin commented:
Can we update docs/reference/commandline/cli.md to mention the new fields in config.json ?

thaJeztah commented:
@duglin we're not in docs review yet, but :+1:

duglin commented:
if i don't say it now i'll forget later  :-)

thaJeztah commented:
> if i don't say it now i'll forget later :-)

Yup, I know the feeling :smile:

jfrazelle commented:
needs rebase :(

vdemeester commented:
@jfrazelle and it's an ugly one :sweat_smile: (that's a nice gift for getting back from sickness :stuck_out_tongue_closed_eyes: ).

vdemeester commented:
And rebased.. hopefully it still builds :sweat_smile:

thaJeztah commented:
fingers crossed!

icecrime commented:
PTAL @tiborvass @cpuguy83!

aaronlehmann commented:
Needs another rebase :hamster:

aaronlehmann commented on daemon/container.go:
```diff
@@ -437,21 +437,27 @@ func attach(streamConfig *runconfig.StreamConfig, openStdin, stdinOnce, tty bool
-					return 0, nil
+					nr, er = src.Read(buf)
+				} else {
+					break
```
How about

```go
if nr != 1 || buf[0] != key {
    break
}
```

and unindent the other block.

vdemeester replied:
Definitely :wink:

aaronlehmann commented on daemon/exec.go:
```diff
@@ -87,6 +88,14 @@ func (d *Daemon) ContainerExecCreate(config *runconfig.ExecConfig) (string, erro
+	if config.DetachKeys != "" {
+		keys, err = term.ToBytes(config.DetachKeys)
+		if err != nil {
+			logrus.Warnf("Wrong escape keys provided (%s) using default : ctrl-p ctrl-q", config.DetachKeys)
```
What does it mean for escape keys to be wrong? Can this provide more detail on what's wrong with them?

vdemeester replied:
I'll add the err string as it says that the espace sequence provided is not supported.

aaronlehmann commented on man/docker-exec.1.md:
```diff
@@ -28,6 +29,9 @@ container is unpaused, and then run
    Detached mode: run command in the background. The default is *false*.
 
+**--detach-keys**=[]
+   Define the key sequence to use to detach from the container.
```
"Define the key sequence which detaches the container"?

MHBauer replied:
I'd add 'from' in there. It's the client doing the detaching, not the container, as I understand.

"Define the key sequence which detaches from the container"?

aaronlehmann replied:
Using `from` seems awkward without saying what's being detached from what. In "Define the key sequence which detaches from the container", it sounds like the key sequence is being detached from the container.

aaronlehmann commented on man/docker-run.1.md:
```diff
@@ -186,7 +187,12 @@ detached container with **docker attach**. If you choose to run a container in
+CTRL-Q. Those keys can be customized using the docker configuration or
+the **--detach-keys** option.
+
+**--detach-eys**=[]
```
Typo - should be detach-keys

aaronlehmann commented on runconfig/exec.go:
```diff
@@ -15,6 +15,7 @@ type ExecConfig struct {
 	AttachStderr bool     // Attach the standard output
 	AttachStdout bool     // Attach the standard error
 	Detach       bool     // Execute in detach mode
+	DetachKeys   string   // Escape keys for unattach
```
s/unattach/detach/

aaronlehmann commented on man/docker-attach.1.md:
```diff
@@ -18,15 +19,19 @@ interactively.  You can attach to the same contained process multiple times
-exit code will be returned to the client.
+You can detach from the container (and leave it running) with the
+escape key sequence, default to `CTRL-p CTRL-q`, (for a quiet exit) or
+`CTRL-c` which will send a `SIGINT` to the container. When you are
```
This sentence is long and a bit awkward. How about:

```
To stop a container, use `CTRL-c`. This key sequence sends `SIGKILL` to the
container. You can detach from the container (and leave it running) using a
configurable key sequence. The default sequence is `CTRL-p CTRL-q`.
```

vdemeester commented:
Rebased and updated :wink:

thaJeztah commented:
Thanks @vdemeester!

vdemeester commented:
And `cli.md` updated @duglin :wink:

duglin commented:
@vdemeester thanks!

vdemeester commented:
rebased :sweat_smile:

thaJeztah commented:
@vdemeester tests are failing

vdemeester commented:
ah damned.. I thought the rebase was too easy to be true :sweat:

calavera commented:
new client implementation looks good @vdemeester! :+1: :zap: 

LGTM

thaJeztah commented:
ping @cpuguy83 can you have another look and move to "docs review" if it LGTY?

thaJeztah commented:
(or @duglin :innocent: )

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -135,6 +136,255 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
+	}
+
+	go func() {
+		exec.Command(dockerBinary, "kill", name).Run()
```
I think we need to remove this.
Detaching should be ending the command shouldn't it?
Or is it stuck on waiting for the container to stop?

Maybe better to use attach directly instead of `docker run`.
This kill command invalidates the detatch test

vdemeester replied:
@cpuguy83 yeah I duplicate the current test but I think it would be better to implement it using the `attach` command ; and doing something like this:

```bash
$ docker run -itd // …
$ docker attach --detach-keys='ctrl-a,a' // …
// Issuing the detach keys sequence
// and assert that the attach command has exited
```

I'll work on rewriting those tests :wink:

cpuguy83 replied:
:heart:

vdemeester commented:
@cpuguy83 done, hopefully should look better :stuck_out_tongue_closed_eyes:

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -120,21 +124,248 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
 	select {
 	case <-ch:
-	case <-time.After(10 * time.Millisecond):
+	case <-time.After(100 * time.Millisecond):
```
Sorry, this timeout could still stand to be much longer... thinking like 10 seconds.
All we care is that it exits and don't want false negatives here, and some extra time on a failing case is OK.

vdemeester replied:
true, I can put 10s :wink:

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -120,21 +124,248 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
+	name := "attach-detach"
+	dockerCmd(c, "run", "--name", name, "-itd", "busybox", "cat")
+
+	cmd := exec.Command(dockerBinary, "attach", "--detach-keys='ctrl-a,a'", name)
```
shouldn't this be handled by the config?

vdemeester replied:
Lol yeah >_<.. Was to quick, I'll remove the flag :stuck_out_tongue_closed_eyes:

vdemeester replied:
hum, this one is fine, I'm making sure the flag override the client config. It's the test above that was wrong >_<.

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -120,21 +124,248 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
+
+	select {
+	case <-ch:
+	case <-time.After(100 * time.Millisecond):
```
same comment as above about the timeout period.

cpuguy83 commented on api/server/router/container/container_routes.go:
```diff
@@ -447,15 +459,27 @@ func (s *containerRouter) wsContainersAttach(ctx context.Context, w http.Respons
+		keys, err = term.ToBytes(detachKeys)
+		if err != nil {
+			logrus.Warnf("Invalid escape keys provided (%s) using default : ctrl-p ctrl-q", detachKeys)
+			// return err
```
can you remove commented code?

cpuguy83 commented on daemon/exec.go:
```diff
@@ -88,6 +89,15 @@ func (d *Daemon) ContainerExecCreate(config *runconfig.ExecConfig) (string, erro
+	if config.DetachKeys != "" {
+		keys, err = term.ToBytes(config.DetachKeys)
+		if err != nil {
+			logrus.Warnf("%s", err.Error())
```
`logrus.Warn(err)`, but probably more helpful to wrap this with some detail.

vdemeester replied:
I put two `Warnf`, not sure if it's needed. Maybe I should merge them. `err` will be a message like "Unknown character: …".

thaJeztah commented:
@vdemeester ouch, looks like this needs a rebase

vdemeester commented:
@thaJeztah yep, I just saw that :sweat_smile: (one less, one more :stuck_out_tongue_closed_eyes: )

vdemeester commented:
Rebased :wink:

cpuguy83 commented:
LGTM, moving to docs

cpuguy83 commented:
On the docs, we should emphasize that things like `ctrl-p,q` is not the same as `ctrl-p,ctrl-q`, because I know we'll get bug reports about it.

Just for fun, I did `--detach-keys e,x,i,t`, and it totally worked nice!
Maybe in the docs make sure to mention it can be more than 2 key presses.

cpuguy83 commented:
Oh, also just noticed if the 2nd key is not escapable, there's no getting back the first key press.
Maybe not a problem, but figured I'd mention.

duglin commented:
Per a chat with @cpuguy83 on irc, we don't send the buffered chars if the user doesn't complete the sequence.  This seems a bit odd to me.  If, once they leave the sequence, the previous chars are lost (meaning they're not sent to the process) then I don't see any point in any sequence that's more than one char. I would think that people want more than one char so that it doesn't overlap with some valid characters they might normally type for some other purpose - and the longer they make it the safer it is. However, once they start down the path of the sequence at all they're locked into typing the entire thing, or losing it all - which makes any prefix of the sequence off limits to the container - which means even just a prefix of one.  So, I'm back to any sequence more than one char doesn't make sense if we don't send the buffer upon leaving the sequence.

the "exit" example is an interesting one (even though its funky).  It means that if I want to type "exiZ" I would need to type "eexiZ", where it seems to me there's no reason why they shouldn't be able to just type "exiZ" w/o worrying about escaping at all since after all "exiZ" isn't the escape sequence.

duglin commented:
However, as @cpuguy83 pointed out, this situation isn't new so technically we could deal with it in a separate PR if wanted - but if we did want to change the semantics it might be nice to do it all under this one.

cpuguy83 commented:
I don't want to change that behavior. Documenting that people should choose a good leader key to prevent this I think is good enough.

thaJeztah commented on docs/reference/commandline/cli.md:
```diff
@@ -100,7 +100,15 @@ The property `psFormat` specifies the default format for `docker ps` output.
+with the `docker attach`, `docker exec`, `docker run` or `docker
+start`, Docker's client uses this property. If this property is not
+set, the client falls back to the default sequence `ctrl-p,ctrl-q`.
+
```
Hm, perhaps this becomes confusing, because the default `CTRL-p CTRL-q` (uppercase CTRL, and no comma) must be specified as `ctrl-p,q` in the configuration?

Are other meta-keys, e.g. `ALT` also possible?

vdemeester replied:
> Hm, perhaps this becomes confusing, because the default CTRL-p CTRL-q (uppercase CTRL, and no comma) must be specified as ctrl-p,q in the configuration?

I'm note sure to understand. The default `CTRL-p CTRL-q` would be `ctrl-p,ctrl-q` in the configuration (or in the `--detach-keys` flag).

> Are other meta-keys, e.g. ALT also possible?

For now, nope :sweat_smile: Could try to make it work though, wdyt ?

thaJeztah commented on man/docker-attach.1.md:
```diff
@@ -18,15 +19,17 @@ interactively.  You can attach to the same contained process multiple times
-exit code will be returned to the client.
+To stop a container, use `CTRL-c`. This key sequence sends `SIGKILL` to the
+container. You can detach from the container (and leave it running) using a
+configurable key sequence. The default sequence is `CTRL-p CTRL-q`.
```
```Markdown
To stop a container, use `CTRL-c`. This key sequence sends `SIGKILL` to the
 container. You can detach from the container (and leave it running) using a
 configurable key sequence. You configure the key sequence using the
`--detach-keys` option. The default sequence is `CTRL-p CTRL-q`, and
can be changed using a configuration file. See **docker(1)** for documentation
on using a config file.
```

Hm, only it looks like the long description of "config files" is not in the man page :/

vdemeester replied:
Should I update the `docker` man page with a config file thinky ? :stuck_out_tongue:

thaJeztah commented on man/docker-attach.1.md:
```diff
@@ -18,15 +19,17 @@ interactively.  You can attach to the same contained process multiple times
 attaching to a tty-enabled container (i.e.: launched with `-t`).
 
 # OPTIONS
+**--detach-keys**=[]
```
Should the `[]` be there? Can you specify the `--detach-keys` option multiple times?

vdemeester replied:
Hum right, should not be `[]` :stuck_out_tongue:

thaJeztah replied:
I suspect that needs a fix then, yes

thaJeztah replied:
oh, actually, that's from the synopsis, I think that's a different format. For example, this is how the `--log-driver` is mentioned under "options":

    **--log-driver**="json-file|syslog|journald|gelf|fluentd|awslogs|splunk|none"

thaJeztah commented on man/docker-attach.1.md:
```diff
@@ -6,6 +6,7 @@ docker-attach - Attach to a running container
 
 # SYNOPSIS
 **docker attach**
+[**--detach-keys**[=*[]*]]
```
Same here (`[]`) - and other places

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -189,7 +190,12 @@ detached container with **docker attach**. If you choose to run a container in
+stopping the process by pressing the detach keys, by default CTRL-P
+CTRL-Q. Those keys can be customized using the docker configuration or
+the **--detach-keys** option.
+
```
Perhaps use the same here?

```Markdown
    When attached in the tty mode, you can detach from the container (and leave it
running) using a configurable key sequence. The default sequence is `CTRL-p CTRL-q`.
You configure the key sequence using the **--detach-keys** option or a configuration file.
See **docker(1)** for documentation on using a configuration file.
```

thaJeztah commented:
Looks like the online documentation for attach, exec and start also need to be updated. Also, can this option be set for *create* and *run*?

vdemeester commented:
> Looks like the online documentation for attach, exec and start also need to be updated. Also, can this option be set for create and run?

Ohh weirdly enough I though I had done these files (the `reference` one, but I guess I lost my changes somewhere :sweat:). I'll update it quicly.

The option can be set for : `attach`, `exec`, `run` and `start`, but not `create` :stuck_out_tongue:.

vdemeester commented:
Update man pages and rebased against master :stuck_out_tongue_closed_eyes:.

thaJeztah commented on man/docker.1.md:
```diff
@@ -223,6 +223,71 @@ inside it)
+Docker's client uses this property. If this property is not set, the client
+falls back to the default table format. For a list of supported formatting
+directives, see the
+[**Formatting** section in the `docker ps` documentation](ps.md)
```
```
see **docker-ps(1)**.
```

vdemeester replied:
damn, you're right :smile_cat:

thaJeztah commented on man/docker.1.md:
```diff
@@ -223,6 +223,71 @@ inside it)
+When the `--format` flag is not provided with the `docker images` command,
+Docker's client uses this property. If this property is not set, the client
+falls back to the default table format. For a list of supported formatting
+directives, see the [**Formatting** section in the `docker images` documentation](images.md)
```
```
see **docker-images(1)**.
```

thaJeztah commented on api/client/run.go:
```diff
@@ -74,6 +74,7 @@ func (cli *DockerCli) CmdRun(args ...string) error {
 		flDetach     = cmd.Bool([]string{"d", "-detach"}, false, "Run container in background and print container ID")
 		flSigProxy   = cmd.Bool([]string{"-sig-proxy"}, true, "Proxy received signals to the process")
 		flName       = cmd.String([]string{"-name"}, "", "Assign a name to the container")
+		flDetachKeys = cmd.String([]string{"-detach-keys"}, "", "Set up escape key sequence")
```
Still not sure if "Set up escape key sequence" is the best wording, but no really good alternative ("Set escape key sequence for detach",  "Set detach key sequence"?

@moxiegirl perhaps you have some suggestions?

moxiegirl replied:
`Specify the escape key sequence used to detach a container`

moxiegirl commented:
I'm reviewing this..please hold the merge I'll be out with the review today.

thaJeztah commented:
@moxiegirl thanks! I also discussed with @vdemeester that docs for the API changes are missing (this also affects the API)

moxiegirl commented:
@vdemeester Can you provide me with some kind of understanding of the format for the value passed for the key. For example, is `ESC-6` possible or can I only do `CTRL-f` and not `^-f` in specifying the key ?

vdemeester commented:
@moxiegirl yes ! For now it's only alphabet keys (a-z) and ctrl key (here : https://github.com/docker/docker/pull/15666/files#diff-e9bc5b8f6d0393981a0f82d17f51da29R9). `ESC-6` or `alt-t` would not work.

moxiegirl commented on docs/reference/commandline/attach.md:
```diff
@@ -14,6 +14,7 @@ parent = "smn_cli"
 
     Attach to a running container
 
+      --detach-keys       Specify the escape key sequence used to detach a container
```
@vdemeester Nice feature. I think this needs a bit of rearranging and more info on the flag: 

---> https://gist.github.com/moxiegirl/b20a296942ab31d62e56

# attach

    Usage: docker attach [OPTIONS] CONTAINER

    Attach to a running container

      --detach-keys="ctrl-<value>"       Set up escape key sequence
      --help              Print usage
      --no-stdin          Do not attach STDIN
      --sig-proxy=true    Proxy all received signals to the process

The `docker attach` command allows you to attach to a running container using
the container's ID or name, either to view its ongoing output or to control it
interactively. You can attach to the same contained process multiple times
simultaneously, screen sharing style, or quickly view the progress of your
detached  process.

To stop a container, use `CTRL-c`. This key sequence sends `SIGKILL` to the
container. If `--sig-proxy` is true (the default),`CTRL-c` sends a `SIGINT` to
the container. You can detach from a container and leave it running using the
using `CTRL-p CTRL-q` key sequence.

> **Note:**
> A process running as PID 1 inside a container is treated specially by
> Linux: it ignores any signal with the default action. So, the process
> will not terminate on `SIGINT` or `SIGTERM` unless it is coded to do
> so.

It is forbidden to redirect the standard input of a `docker attach` command
while attaching to a tty-enabled container (i.e.: launched with `-t`).

## Override the detach sequence

If you want, you can configure a different sequence for detach. Configuration is
useful if the Docker default sequence overrides a default you use for other
applications. There are two ways to configure a custom detach key sequence, per
container or on your entire configuration.

To change the sequence for a particular container, use the
`--detach-keys="ctrl-<value>"` flag with the `docker attach` command. Replace
`<value>` with any of the following characters:

* `a-z` (a single lowercase alpha character )
* `@` (ampersand)
* `[` (left bracket)
* `\\` (two backward slashes)
*  `_` (underscore)
* `^` (caret)

To configure a different key sequence for all containers, see [**Configuration
file** section](cli.md#configuration-files).

thaJeztah replied:
@moxiegirl I think the text should be more explicit that the sequence is a comma-separated list. E.g. `ctrl-a,e’, `ctrl-a,ctrl-b`, `ctrl-a,b,c,ctrl-d` are all valid (I think). The first keystroke doesn't have to be `ctrl-...` (although using a regular character is usually impractical)

moxiegirl replied:
Hmm, the way I read @vdemeester comment the `ctrl-` was hardcoded but the follow on value was not. That's what the code looks like. Let's see what he says, I can tweak the language.

vdemeester replied:
@moxiegirl ah ok my comment wasn't as good as it could then ; @thaJeztah is right, it can be either letter or `ctrl-…` where `…` is a letter. :stuck_out_tongue_closed_eyes:

moxiegirl replied:
@vdemeester No worries. Thanks for clarifying, then...I think this:

```markdown
To override the sequence for an individual container, use the
--detach-keys="<sequence>" flag with the `docker attach` command. The format of
the `<sequence>` is either a letter [a-Z], or the `ctrl-` combined with any of
the following:


a-z (a single lowercase alpha character )
@ (ampersand)
[ (left bracket)
\\ (two backward slashes)
_ (underscore)
^ (caret)

These `a`, `ctrl-a`, `X`, or `ctrl-\\` values are all examples of valid key
sequences. To configure a different configuration default key sequence for all
containers, see [**Configuration file** section](cli.md#configuration-files).
```

You'll need to update the minor  sections also: 

```markdown
Override the key sequence for detaching a container. Format is a single character [a-Z] or ctrl-<value> where <value> is one of: a-z, @, ^, [,, or _
```

moxiegirl commented on daemon/exec.go:
```diff
@@ -88,6 +89,14 @@ func (d *Daemon) ContainerExecCreate(config *types.ExecConfig) (string, error) {
+	if config.DetachKeys != "" {
+		keys, err = term.ToBytes(config.DetachKeys)
+		if err != nil {
+			logrus.Warnf("Wrong escape keys provided (%s, error: %s) using default : ctrl-p ctrl-q", config.DetachKeys, err.Error())
```
@vdemeester When does a user get this? Looks like this happens if the configuration is bad.  In which case:

"You specified an unsupported escape key format (%s, error: %s) using the default: ctrl-p ctrl-q"

vdemeester replied:
@moxiegirl the use gets this when he specified a wrong key sequence (either in the config file or using flags)

moxiegirl commented on api/client/attach.go:
```diff
@@ -18,6 +18,7 @@ func (cli *DockerCli) CmdAttach(args ...string) error {
 	cmd := Cli.Subcmd("attach", []string{"CONTAINER"}, Cli.DockerCommands["attach"].Description, true)
 	noStdin := cmd.Bool([]string{"-no-stdin"}, false, "Do not attach STDIN")
 	proxy := cmd.Bool([]string{"-sig-proxy"}, true, "Proxy all received signals to the process")
+	detachKeys := cmd.String([]string{"-detach-keys"}, "", "Specify the escape key sequence used to detach a container")
```
`Specify a custom key sequence for detaching a container.`

moxiegirl commented on api/client/exec.go:
```diff
@@ -16,6 +16,7 @@ import (
 // Usage: docker exec [OPTIONS] CONTAINER COMMAND [ARG...]
 func (cli *DockerCli) CmdExec(args ...string) error {
 	cmd := Cli.Subcmd("exec", []string{"CONTAINER COMMAND [ARG...]"}, Cli.DockerCommands["exec"].Description, true)
+	detachKeys := cmd.String([]string{"-detach-keys"}, "", "Specify the escape key sequence used to detach a container")
```
`Specify a custom key sequence for detaching a container.`

moxiegirl commented on api/client/run.go:
```diff
@@ -74,6 +74,7 @@ func (cli *DockerCli) CmdRun(args ...string) error {
 		flDetach     = cmd.Bool([]string{"d", "-detach"}, false, "Run container in background and print container ID")
 		flSigProxy   = cmd.Bool([]string{"-sig-proxy"}, true, "Proxy received signals to the process")
 		flName       = cmd.String([]string{"-name"}, "", "Assign a name to the container")
+		flDetachKeys = cmd.String([]string{"-detach-keys"}, "", "Specify the escape key sequence used to detach a container")
```
`Specify a custom key sequence for detaching a container.`

moxiegirl commented on api/client/start.go:
```diff
@@ -49,6 +49,7 @@ func (cli *DockerCli) CmdStart(args ...string) error {
 	cmd := Cli.Subcmd("start", []string{"CONTAINER [CONTAINER...]"}, Cli.DockerCommands["start"].Description, true)
 	attach := cmd.Bool([]string{"a", "-attach"}, false, "Attach STDOUT/STDERR and forward signals")
 	openStdin := cmd.Bool([]string{"i", "-interactive"}, false, "Attach container's STDIN")
+	detachKeys := cmd.String([]string{"-detach-keys"}, "", "Specify the escape key sequence used to detach a container")
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

vdemeester replied:
Hum it appears it's too long for one line and it won't allow me multi-line option "comment" (see DockerSuite.TestHelpTextVerify). So… I don't know what to put there :stuck_out_tongue_closed_eyes:

duglin replied:
How about "escape key sequence to detach"?

-Doug

Sent from my iPhone

> On Jan 2, 2016, at 11:53 AM, Vincent Demeester <notifications@github.com> wrote:
> 
> In api/client/start.go:
> 
> > @@ -49,6 +49,7 @@ func (cli *DockerCli) CmdStart(args ...string) error {
> >  	cmd := Cli.Subcmd("start", []string{"CONTAINER [CONTAINER...]"}, Cli.DockerCommands["start"].Description, true)
> >  	attach := cmd.Bool([]string{"a", "-attach"}, false, "Attach STDOUT/STDERR and forward signals")
> >  	openStdin := cmd.Bool([]string{"i", "-interactive"}, false, "Attach container's STDIN")
> > +	detachKeys := cmd.String([]string{"-detach-keys"}, "", "Specify the escape key sequence used to detach a container")
> Hum it appears it's too long for one line and it won't allow me multi-line option "comment" (see DockerSuite.TestHelpTextVerify). So… I don't know what to put there 
> 
> —
> Reply to this email directly or view it on GitHub.
>

moxiegirl commented on api/server/router/container/container_routes.go:
```diff
@@ -450,15 +462,26 @@ func (s *containerRouter) wsContainersAttach(ctx context.Context, w http.Respons
+	if detachKeys != "" {
+		keys, err = term.ToBytes(detachKeys)
+		if err != nil {
+			logrus.Warnf("Invalid escape keys provided (%s) using default : ctrl-p ctrl-q", detachKeys)
```
"Your configuration specifies an unsupported escape key format (%s) using the default: ctrl-p ctrl-q"

moxiegirl commented on docs/reference/api/docker_remote_api_v1.22.md:
```diff
@@ -922,6 +921,10 @@ Start the container `id`
 
+Query Parameters:
+
+-   **detacheys** – Specify the escape key sequence used to detach a container.
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on docs/reference/api/docker_remote_api_v1.22.md:
```diff
@@ -1133,6 +1136,7 @@ Attach to the container `id`
 
 Query Parameters:
 
+-   **detacheys** – Specify the escape key sequence used to detach a container.
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on docs/reference/api/docker_remote_api_v1.22.md:
```diff
@@ -1213,6 +1217,7 @@ Implements websocket protocol handshake according to [RFC 6455](http://tools.iet
 
 Query Parameters:
 
+-   **detacheys** – Specify the escape key sequence used to detach a container.
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on docs/reference/api/docker_remote_api_v1.22.md:
```diff
@@ -2441,6 +2447,7 @@ Json Parameters:
 -   **AttachStdin** - Boolean value, attaches to `stdin` of the `exec` command.
 -   **AttachStdout** - Boolean value, attaches to `stdout` of the `exec` command.
 -   **AttachStderr** - Boolean value, attaches to `stderr` of the `exec` command.
+-   **Detacheys** – Specify the escape key sequence used to detach a container.
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on docs/reference/commandline/cli.md:
```diff
@@ -101,7 +101,15 @@ The property `psFormat` specifies the default format for `docker ps` output.
 Docker's client uses this property. If this property is not set, the client
 falls back to the default table format. For a list of supported formatting
-directives, see the [**Formatting** section in the `docker ps` documentation](ps.md)
+directives, see the
```
```markdown
Once attached to a container, users detach from it and leave it running using the using `CTRL-p CTRL-q` key sequence. This detach key sequence is customizable  using the `detachKeys` property. Specify a `ctrl-<value>` value for the property. Replace the `<value>` component with any of the following characters:

* `a-z` (a single lowercase alpha character )
* `@` (ampersand)
* `[` (left bracket)
* `\\` (two backward slashes)
*  `_` (underscore)
* `^` (caret)

Your customization applies to all containers started in with your Docker client. Users can override your custom or the default key sequence on a per-container basis. To do this, the user specifies the `--detach-keys` flag with the `docker attach`, `docker exec`, `docker run` or `docker start` command. 
```

moxiegirl commented on man/docker-attach.1.md:
```diff
@@ -18,15 +19,19 @@ interactively.  You can attach to the same contained process multiple times
-(for a quiet exit) or `CTRL-c` which will send a `SIGKILL` to the container.
-When you are attached to a container, and exit its main process, the process's
-exit code will be returned to the client.
+To stop a container, use `CTRL-c`. This key sequence sends `SIGKILL` to the
```
Use the content from the `attach.md` file provided above.

moxiegirl commented on man/docker-attach.1.md:
```diff
@@ -18,15 +19,19 @@ interactively.  You can attach to the same contained process multiple times
 
 # OPTIONS
+**--detach-keys**=""
+  Define the key sequence which detaches the container.
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on man/docker-exec.1.md:
```diff
@@ -28,6 +29,9 @@ container is unpaused, and then run
 
+**--detach-keys**=""
+  Define the key sequence which detaches the container.
+
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on man/docker-exec.1.md:
```diff
@@ -7,6 +7,7 @@ docker-exec - Run a command in a running container
 # SYNOPSIS
 **docker exec**
 [**-d**|**--detach**]
+[**--detach-keys**[=*[]*]]
```
@vdemeester is the `=` required or not...I didn't see it on the commands or did I?

vdemeester replied:
@moxiegirl I think it's the same as some other flags, where you can use them with or without the `=` sign.

moxiegirl commented on man/docker-exec.1.md:
```diff
@@ -28,6 +29,9 @@ container is unpaused, and then run
    Detached mode: run command in the background. The default is *false*.
 
+**--detach-keys**=""
+  Define the key sequence which detaches the container.
```
`Specify a custom key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on man/docker-run.1.md:
```diff
@@ -190,8 +191,13 @@ the other shell to view a list of the running containers. You can reattach to a
+See **docker(1)** for documentation on using a configuration file.
+
+**--detach-keys**=""
+   Define the key sequence to use to detach from the container.
```
`Override the key sequence for detaching a container. Format is ctrl-<value> where <value> is one of: a-z, @, ^, [, ``, or _`

moxiegirl commented on man/docker-run.1.md:
```diff
@@ -190,8 +191,13 @@ the other shell to view a list of the running containers. You can reattach to a
 
-   When attached in the tty mode, you can detach from a running container without
-stopping the process by pressing the keys CTRL-P CTRL-Q.
+   When attached in the tty mode, you can detach from the container (and leave it
```
Please use the text from the `attached.md` section on this above.

vdemeester replied:
Hum, I didn't copy it everywhere, true :stuck_out_tongue_closed_eyes:. I just updated this one as it was present :smile_cat:

moxiegirl commented on man/docker.1.md:
```diff
@@ -223,6 +223,70 @@ inside it)
   Block until a container stops, then print its exit code
   See **docker-wait(1)** for full documentation on the **wait** command.
 
+# CONFIGURATION FILES
```
Hmm, do we really want this here? I think a more appropriate thing to do would be to create a man page for the `config.json` file and put this there.

vdemeester replied:
Makes sence to me, but I wonder if I should do this here or in another PR ? :stuck_out_tongue_closed_eyes:

moxiegirl replied:
Another PR. @vdemeester am I volunteering or being volunteered. How about I make a PR and you remove it here?

vdemeester replied:
Perfect :wink: :stuck_out_tongue_closed_eyes:

moxiegirl commented:
@vdemeester Oh, man, I'm sorry to do this review so interactively.  I think the language is:

- create a custom key sequence in `config.json` for all containers
- override the detach key sequence on a specific container with the flag

It took me a bit to get there.

moxiegirl commented:
LGTM for the @vdemeester! :raised_hands:

thaJeztah commented on api/client/attach.go:
```diff
@@ -18,6 +18,7 @@ func (cli *DockerCli) CmdAttach(args ...string) error {
 	cmd := Cli.Subcmd("attach", []string{"CONTAINER"}, Cli.DockerCommands["attach"].Description, true)
 	noStdin := cmd.Bool([]string{"-no-stdin"}, false, "Do not attach STDIN")
 	proxy := cmd.Bool([]string{"-sig-proxy"}, true, "Proxy all received signals to the process")
+	detachKeys := cmd.String([]string{"-detach-keys"}, "", `Override the key sequence for detaching a container`)
```
Should this be regular quotes, not back-tics?

vdemeester replied:
it should but it is no big deal :sweat_smile:.

thaJeztah commented:
@vdemeester One minor question https://github.com/docker/docker/pull/15666#discussion_r48697550, probably no need to update it, but just checking. Otherwise;

LGTM


Thanks @vdemeester, let's get this merged! :dancer: 

(I'll move to "merge", but if you think https://github.com/docker/docker/pull/15666#discussion_r48697550 needs to be updated, feel free to do so, and merge afterwards)

thaJeztah commented:
damn you github, why are you removing/adding the wrong labels :cry:

vdemeester committed:
Implement configurable detach key

Implement configurable detach keys (for `attach`, exec`, `run` and
`start`) using the client-side configuration

- Adds a `--detach-keys` flag to `attach`, `exec`, `run` and `start`
  commands.
- Adds a new configuration field (in `~/.docker/config.json`) to
  configure the default escape keys for docker client.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah commented:
It's green :tada: \o/

vdemeester commented:
:tada: :santa:

pulserdd commented:
Hi, I have a question on using this detachKeys config. We are using ecs instance on which these docker tasks are run. Following the documentation, I added the "detachKeys": "ctrl-e" to /root/.docker/config.json but it does not seem to take effect. I still have to use ctrl-c to exit on attach and that kills the container. Could someone share some pointers on resolving this? I have tried restarting the docker daemon but to no avail. Would highly appreciate some help. Thanks.

dryfish commented:
@pulserdd I've just tested with 1.11.1 and `"DetachKeys": "ctrl-x"` is working for me. I think docker JSON files always use title casing.

thaJeztah commented:
hm, in that case either the documentation is incorrect, or something changed in the parsing of that file https://docs.docker.com/engine/reference/commandline/cli/#configuration-files

@vdemeester ?

dryfish commented:
Ah. I've been looking at docker-machine configs recently and seeing title case.

I've just tried `docker run` with `detachKeys` and that *also* works for me.

vdemeester commented:
@pulserdd tried it out and work as expected. In the config file it takes both `detachKeys` or `DetachKeys`. As @dryfish it has to be set in `$HOME/.docker/config.json` of the user that do the `docker run`, or `docker attach`, or … It's client-side so no need to restart the daemon or anything.

pulserdd commented:
Thanks for your replies. I tried to follow your comments but still the same issue. Here's what I have in config
`[root@ip-1-2-3-4 ec2-user]# cat ~/.docker/config.json 
{
	"auths": {
		"quay.io": {
			"auth": "xxx",
			"email": ""
		}
	},
	"DetachKeys": "ctrl-e"
}`

Result:
`[root@ip-1-2-3-4 ec2-user]# docker attach 7fe7a4ed6213
E E Ee
C2016-05-09 18:06:05,463 INFO [cs.iosmdm.iosmdm] wsgi exiting`

my docker version
`Docker version 1.9.1, build a34a1d5/1.9.1`

vdemeester commented:
@pulserdd oh it's normal then, this feature is only available starting docker 1.10…

pulserdd commented:
oh ok sorry I missed that while following the docs. Thanks once again for a great community support.

MhdSyrwan commented:
@vdemeester Can you please correct the example in the first post since people sometimes refer to this pull request as a documentation page (like me :smile: ).

The wrong example:
```json
{
    "detachKey": "ctrl-a,a"
}
```

The correct one:
```json
{
    "detachKeys": "ctrl-a,a"
}
```

thaJeztah commented:
@MhdSyrwan thanks for letting us know! I updated the description
